### PR TITLE
Reader: Expose stream selected card as a post key, fix up combos

### PR DESCRIPTION
**Pull request by blowery**
This teaches the feed-stream-store to hide the selected index, only exposing the selected post key. It also updates all of the clients of the feed-stream-store (and paged-stream) to use the post key when selecting items.

### CODE REVIEW

**Comment by samouri** [comment]
looking at this now

File: client/reader/stream/index.jsx [hunk 1]
```diff
@@ -220,7 +220,7 @@ class ReaderStream extends React.Component {
 			posts,
 			recs,
 			updateCount: store.getUpdateCount(),
-			selectedIndex: store.getSelectedIndex(),
+			selectedPost: store.getSelectedPost(),
```
Comment (samouri) [inline client/reader/stream/index.jsx]: is this a selectedPost or a selectedPostKey

Comment (blowery) [inline client/reader/stream/index.jsx]: key

File: client/blocks/reader-combined-card/index.jsx [hunk 2]
```diff
@@ -15,13 +15,22 @@ import ReaderSiteStreamLink from 'blocks/reader-site-stream-link';
...
 
-const ReaderCombinedCard = ( { posts, site, feed, onClick, isDiscover, translate } ) => {
+const ReaderCombinedCard = ( { posts, site, feed, selectedPost, onClick, isDiscover, translate } ) => {
 	const feedId = get( feed, 'feed_ID' );
 	const siteId = get( site, 'ID' );
 	const siteIcon = get( site, 'icon.img' );
 	const feedIcon = get( feed, 'image' );
 	const streamUrl = getStreamUrl( feedId, siteId );
 	const siteName = siteNameFromSiteAndPost( site, posts[ 0 ] );
+	const isSelectedPost = post => {
```
Comment (samouri) [inline client/blocks/reader-combined-card/index.jsx]: I think if we created a couple postKey related primitives in reader/util then some of our logic could end up similar

```jsx
postToPostKey( post ) => postKey
combinedCardToPostKeys( cc ) => [ postKeys, ... ] // or maybe combined card shouldn't have list of postIds and should have a list of postKeys instead to make things simpler.  denormalization ftw?
```

then isSelectedPost becomes:

```jsx
const isSelectedPost = post => _.isEqual( postToPostKey( post ), postToPostKey( selectedPost ) );
```

File: client/reader/stream/index.jsx (hunk 1 above)
Comment (samouri) [inline client/reader/stream/index.jsx]: lets change that variable name so its not a lie!

**Comment by samouri** [comment]
other than https://github.com/Automattic/wp-calypso/pull/11965#discussion_r105275715, code looks good. testing now

**Comment by samouri** [comment]
functionality is sweet! nice work. 

The combined cards are missing the styling though on my browser (trust me its selected): 
screen shot 2017-03-09 at 5 04 04 pm

**Comment by blowery** [comment]
@samouri fixed the styling, moved to selectedPostKey, added utils

**Comment by blowery** [comment]
rebased on master too

File: client/lib/feed-stream-store/post-key.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+export function keyForPost( post ) {
+	if ( post.feed_ID && post.feed_item_ID ) {
+		return {
+			feedId: post.feed_ID,
+			postId: post.feed_item_ID
+		};
+	}
+	if ( post.is_external ) {
```
Comment (samouri) [inline client/lib/feed-stream-store/post-key.js]: for my own learning: what does is_external mean?

File: client/lib/feed-stream-store/post-key.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+			postId: post.ID
+		};
+	}
+	return {
+		blogId: post.site_ID,
+		postId: post.ID
+	};
+}
+
+export function keysAreEqual( a, b ) {
```
Comment (samouri) [inline client/lib/feed-stream-store/post-key.js]: whats the benefit of this over _.isEquals() ?

Comment (blowery) [inline client/lib/feed-stream-store/post-key.js]: it doesn't care if there's an undefined extra prop.

`_.isEqual( { feedId: 1, postId: 2 }, { feedId: 1, postId: 2, blogId: undefined } )` is false

File: client/lib/feed-stream-store/post-key.js (hunk 3 above)
Comment (blowery) [inline client/lib/feed-stream-store/post-key.js]: the post came from a feed that's not on jetpack or wpcom

File: client/lib/feed-stream-store/post-key.js (hunk 4 above)
Comment (samouri) [inline client/lib/feed-stream-store/post-key.js]: that makes me think that we should have all our `postKey` generators to end with:
`return pickBy( postKey, identity)`.

But i do really like your keysAreEqual function 🤔

File: client/lib/feed-stream-store/post-key.js (hunk 3 above)
Comment (samouri) [inline client/lib/feed-stream-store/post-key.js]: thanks!

**Comment by samouri** [comment]
Thanks for adding the keyForPost util! I selfishly wanted that for a PR i'm working on :)

File: client/blocks/reader-combined-card/style.scss [hunk 5]
```diff
@@ -82,6 +82,23 @@
 	display: flex;
 	flex-direction: row;
 	margin-bottom: 20px;
+	position: relative;
+
+	&.is-selected {
+		&::before {
+			content: '';
+			position: absolute;
```
Comment (samouri) [inline client/blocks/reader-combined-card/style.scss]: I think this needs to be `relative`! then lets 🚢  this its sweet

Comment (samouri) [inline client/blocks/reader-combined-card/style.scss]: also width needs to b 3px to match the others

**Comment by samouri** [comment]
@blowery, make sure those styles work for you and then feel free to 🚢

File: client/blocks/reader-combined-card/style.scss (hunk 5 above)
Comment (blowery) [inline client/blocks/reader-combined-card/style.scss]: eh? these are pulled directly from the existing styles. https://github.com/Automattic/wp-calypso/blob/master/client/blocks/reader-post-card/style.scss#L37

File: client/lib/feed-stream-store/post-key.js (hunk 3 above)
Comment (blowery) [inline client/lib/feed-stream-store/post-key.js]: right, and it seems they don't actually have a `feed_item_ID` set at all. odd.

**Comment by blowery** [comment]
@samouri those style changes broke it for me. Selected combined posts are now pushed a few pixels to the right and the bar is too thick. It's 2px on the normal cards.

**Comment by blowery** [comment]
_5__following_ _reader_ _wordpress_com

File: client/lib/feed-stream-store/post-key.js (hunk 4 above)
Comment (blowery) [inline client/lib/feed-stream-store/post-key.js]: hmm, i think this is actually busted. doesn't deal with gaps. depending on usage, might need to deal with recs blocks and combos too. I don't really consider those postKeys (even gaps are questionable, but at least they're part of the stream data), but item descriptors... 

one nice thing about isEqual is that it just worked with things like gaps, but it requires that the objects all agree on a shape, which can lead to subtle bugs.

**Comment by blowery** [comment]
@samouri what are you seeing that led you to make the changes you made? It looks perfect on my end without them...

**Comment by samouri** [comment]
@blowery, I remember that until I changed the position I couldn't see any blue bars.  And then they also looked thinner than the regular card blue lines.

But today it works perfectly for me so I've probably gone bonkers

**Pull request merged**